# Fine-tuning a model on a Q&A task using SQUAD

## In this notebook, we will see how to fine-tune one of the [🤗 Transformers](https://github.com/huggingface/transformers) model to a question answering task, which is the task of extracting the answer to a question from a given context. We will see how to easily load a dataset for these kinds of tasks and use the `Trainer` API to fine-tune a model on it.

![Widget inference representing the QA task](images/question_answering.png)

**Note:** This notebook finetunes models that answer question by taking a substring of a context, not by generating new text.

Make sure your version of Transformers is at least 4.8.1 since the functionality was introduced in that version:

In [1]:
import transformers

print(transformers.__version__)

4.9.2


This notebook is built to run on any question answering task with the same format as SQUAD (version 1 or 2), with any model checkpoint from the [Model Hub](https://huggingface.co/models) as long as that model has a version with a token classification head and a fast tokenizer (check on [this table](https://huggingface.co/transformers/index.html#bigtable) if this is the case). It might just need some small adjustments if you decide to use a different dataset than the one used here. Depending on you model and the GPU you are using, you might need to adjust the batch size to avoid out-of-memory errors. Set those three parameters, then the rest of the notebook should run smoothly:

In [2]:
# This flag is the difference between SQUAD v1 or 2 (if you're using another dataset, it indicates if impossible
# answers are allowed or not).
squad_v2 = False
model_checkpoint = "distilbert-base-uncased"
batch_size = 24

## Loading the dataset

We will use the [🤗 Datasets](https://github.com/huggingface/datasets) library to download the data and get the metric we need to use for evaluation (to compare our model to the benchmark). This can be easily done with the functions `load_dataset` and `load_metric`.  

In [3]:
from datasets import load_dataset, load_metric

For our example here, we'll use the [SQUAD dataset](https://rajpurkar.github.io/SQuAD-explorer/). The notebook should work with any question answering dataset provided by the 🤗 Datasets library. If you're using your own dataset defined from a JSON or csv file (see the [Datasets documentation](https://huggingface.co/docs/datasets/loading_datasets.html#from-local-files) on how to load them), it might need some adjustments in the names of the columns used.

In [4]:
datasets = load_dataset("squad_v2" if squad_v2 else "squad")

Reusing dataset squad (/home/john/.cache/huggingface/datasets/squad/plain_text/1.0.0/d6ec3ceb99ca480ce37cdd35555d6cb2511d223b9150cce08a837ef62ffea453)


The `datasets` object itself is [`DatasetDict`](https://huggingface.co/docs/datasets/package_reference/main_classes.html#datasetdict), which contains one key for the training, validation and test set.

In [5]:
datasets

DatasetDict({
    train: Dataset({
        features: ['id', 'title', 'context', 'question', 'answers'],
        num_rows: 87599
    })
    validation: Dataset({
        features: ['id', 'title', 'context', 'question', 'answers'],
        num_rows: 10570
    })
})

We can see the training, validation and test sets all have a column for the context, the question and the answers to those questions.

To access an actual element, you need to select a split first, then give an index:

In [6]:
datasets["train"][0]

{'id': '5733be284776f41900661182',
 'title': 'University_of_Notre_Dame',
 'context': 'Architecturally, the school has a Catholic character. Atop the Main Building\'s gold dome is a golden statue of the Virgin Mary. Immediately in front of the Main Building and facing it, is a copper statue of Christ with arms upraised with the legend "Venite Ad Me Omnes". Next to the Main Building is the Basilica of the Sacred Heart. Immediately behind the basilica is the Grotto, a Marian place of prayer and reflection. It is a replica of the grotto at Lourdes, France where the Virgin Mary reputedly appeared to Saint Bernadette Soubirous in 1858. At the end of the main drive (and in a direct line that connects through 3 statues and the Gold Dome), is a simple, modern stone statue of Mary.',
 'question': 'To whom did the Virgin Mary allegedly appear in 1858 in Lourdes France?',
 'answers': {'text': ['Saint Bernadette Soubirous'], 'answer_start': [515]}}

We can see the answers are indicated by their start position in the text (here at character 515) and their full text, which is a substring of the context as we mentioned above.

To get a sense of what the data looks like, the following function will show some examples picked randomly in the dataset (automatically decoding the labels in passing).

In [7]:
from datasets import ClassLabel, Sequence
import random
import pandas as pd
from IPython.display import display, HTML

def show_random_elements(dataset, num_examples=5):
    assert num_examples <= len(dataset), "Can't pick more elements than there are in the dataset."
    picks = []
    for _ in range(num_examples):
        pick = random.randint(0, len(dataset)-1)
        while pick in picks:
            pick = random.randint(0, len(dataset)-1)
        picks.append(pick)
    
    df = pd.DataFrame(dataset[picks])
    for column, typ in dataset.features.items():
        if isinstance(typ, ClassLabel):
            df[column] = df[column].transform(lambda i: typ.names[i])
        elif isinstance(typ, Sequence) and isinstance(typ.feature, ClassLabel):
            df[column] = df[column].transform(lambda x: [typ.feature.names[i] for i in x])
    display(HTML(df.to_html()))

In [8]:
show_random_elements(datasets["train"])

## Preprocessing the training data

Before we can feed those texts to our model, we need to preprocess them. This is done by a 🤗 Transformers `Tokenizer` which will (as the name indicates) tokenize the inputs (including converting the tokens to their corresponding IDs in the pretrained vocabulary) and put it in a format the model expects, as well as generate the other inputs that model requires.

To do all of this, we instantiate our tokenizer with the `AutoTokenizer.from_pretrained` method, which will ensure:

- we get a tokenizer that corresponds to the model architecture we want to use,
- we download the vocabulary used when pretraining this specific checkpoint.

That vocabulary will be cached, so it's not downloaded again the next time we run the cell.

In [9]:
from transformers import AutoTokenizer
    
tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)

The following assertion ensures that our tokenizer is a fast tokenizers (backed by Rust) from the 🤗 Tokenizers library. Those fast tokenizers are available for almost all models, and we will need some of the special features they have for our preprocessing.

In [10]:
import transformers
assert isinstance(tokenizer, transformers.PreTrainedTokenizerFast)

You can check which type of models have a fast tokenizer available and which don't on the [big table of models](https://huggingface.co/transformers/index.html#bigtable).

You can directly call this tokenizer on two sentences (one for the answer, one for the context):

In [11]:
tokenizer("What is your name?", "My name is Sylvain.")

{'input_ids': [101, 2054, 2003, 2115, 2171, 1029, 102, 2026, 2171, 2003, 25353, 22144, 2378, 1012, 102], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}

Depending on the model you selected, you will see different keys in the dictionary returned by the cell above. They don't matter much for what we're doing here (just know they are required by the model we will instantiate later), you can learn more about them in [this tutorial](https://huggingface.co/transformers/preprocessing.html) if you're interested.

Now one specific thing for the preprocessing in question answering is how to deal with very long documents. We usually truncate them in other tasks, when they are longer than the model maximum sentence length, but here, removing part of the the context might result in losing the answer we are looking for. To deal with this, we will allow one (long) example in our dataset to give several input features, each of length shorter than the maximum length of the model (or the one we set as a hyper-parameter). Also, just in case the answer lies at the point we split a long context, we allow some overlap between the features we generate controlled by the hyper-parameter `doc_stride`:

In [12]:
max_length = 384 # The maximum length of a feature (question and context)
doc_stride = 128 # The authorized overlap between two part of the context when splitting it is needed.

Let's find one long example in our dataset:

In [13]:
for i, example in enumerate(datasets["train"]):
    if len(tokenizer(example["question"], example["context"])["input_ids"]) > max_length:
        break
print(f"index {i} will span multiple features.\n\n sample = {datasets['train'][i]} ")

index 249 will span multiple features.

 sample = {'id': '5733caf74776f4190066124c', 'title': 'University_of_Notre_Dame', 'context': "The men's basketball team has over 1,600 wins, one of only 12 schools who have reached that mark, and have appeared in 28 NCAA tournaments. Former player Austin Carr holds the record for most points scored in a single game of the tournament with 61. Although the team has never won the NCAA Tournament, they were named by the Helms Athletic Foundation as national champions twice. The team has orchestrated a number of upsets of number one ranked teams, the most notable of which was ending UCLA's record 88-game winning streak in 1974. The team has beaten an additional eight number-one teams, and those nine wins rank second, to UCLA's 10, all-time in wins against the top team. The team plays in newly renovated Purcell Pavilion (within the Edmund P. Joyce Center), which reopened for the beginning of the 2009–2010 season. The team is coached by Mike Brey, who, 

Without any truncation, we get the following length for the input IDs:

In [14]:
len(tokenizer(example["question"], example["context"])["input_ids"])

396

Now, if we just truncate, we will lose information (and possibly the answer to our question):

In [15]:
len(tokenizer(example["question"], example["context"], max_length=max_length, truncation="only_second")["input_ids"])

384

Note that we never want to truncate the question, only the context, else the `only_second` truncation picked. Now, our tokenizer can automatically return us a list of features capped by a certain maximum length, with the overlap we talked above, we just have to tell it with `return_overflowing_tokens=True` and by passing the stride:

In [16]:
tokenized_example = tokenizer(
    example["question"],
    example["context"],
    max_length=max_length,
    truncation="only_second",
    return_overflowing_tokens=True,
    stride=doc_stride
)

In [17]:
len(tokenized_example['input_ids'][0])

384

Now we don't have one list of `input_ids`, but several: 

In [18]:
[len(x) for x in tokenized_example["input_ids"]]

[384, 157]

And if we decode them, we can see the overlap:

In [19]:
for x in tokenized_example["input_ids"][:2]:
    print(f"{tokenizer.decode(x)}\n" )

[CLS] how many wins does the notre dame men's basketball team have? [SEP] the men's basketball team has over 1, 600 wins, one of only 12 schools who have reached that mark, and have appeared in 28 ncaa tournaments. former player austin carr holds the record for most points scored in a single game of the tournament with 61. although the team has never won the ncaa tournament, they were named by the helms athletic foundation as national champions twice. the team has orchestrated a number of upsets of number one ranked teams, the most notable of which was ending ucla's record 88 - game winning streak in 1974. the team has beaten an additional eight number - one teams, and those nine wins rank second, to ucla's 10, all - time in wins against the top team. the team plays in newly renovated purcell pavilion ( within the edmund p. joyce center ), which reopened for the beginning of the 2009 – 2010 season. the team is coached by mike brey, who, as of the 2014 – 15 season, his fifteenth at notr

Now this will give us some work to properly treat the answers: we need to find in which of those features the answer actually is, and where exactly in that feature. The models we will use require the start and end positions of these answers in the tokens, so we will also need to to map parts of the original context to some tokens. Thankfully, the tokenizer we're using can help us with that by returning an `offset_mapping`:

In [20]:
tokenized_example = tokenizer(
    example["question"],
    example["context"],
    max_length=max_length,
    truncation="only_second",
    return_overflowing_tokens=True,
    return_offsets_mapping=True,
    stride=doc_stride
)
print(tokenized_example["offset_mapping"][0][:100])

[(0, 0), (0, 3), (4, 8), (9, 13), (14, 18), (19, 22), (23, 28), (29, 33), (34, 37), (37, 38), (38, 39), (40, 50), (51, 55), (56, 60), (60, 61), (0, 0), (0, 3), (4, 7), (7, 8), (8, 9), (10, 20), (21, 25), (26, 29), (30, 34), (35, 36), (36, 37), (37, 40), (41, 45), (45, 46), (47, 50), (51, 53), (54, 58), (59, 61), (62, 69), (70, 73), (74, 78), (79, 86), (87, 91), (92, 96), (96, 97), (98, 101), (102, 106), (107, 115), (116, 118), (119, 121), (122, 126), (127, 138), (138, 139), (140, 146), (147, 153), (154, 160), (161, 165), (166, 171), (172, 175), (176, 182), (183, 186), (187, 191), (192, 198), (199, 205), (206, 208), (209, 210), (211, 217), (218, 222), (223, 225), (226, 229), (230, 240), (241, 245), (246, 248), (248, 249), (250, 258), (259, 262), (263, 267), (268, 271), (272, 277), (278, 281), (282, 285), (286, 290), (291, 301), (301, 302), (303, 307), (308, 312), (313, 318), (319, 321), (322, 325), (326, 330), (330, 331), (332, 340), (341, 351), (352, 354), (355, 363), (364, 373), (374,

In [21]:
len(tokenized_example['offset_mapping'][1])

157

This gives, for each index of our input IDS, the corresponding start and end character in the original text that gave our token. The very first token (`[CLS]`) has (0, 0) because it doesn't correspond to any part of the question/answer, then the second token is the same as the characters 0 to 3 of the question:

In [22]:
first_token_id = tokenized_example["input_ids"][0][1]
offsets = tokenized_example["offset_mapping"][0][1]
print(tokenizer.convert_ids_to_tokens([first_token_id])[0], example["question"][offsets[0]:offsets[1]])

how How


So we can use this mapping to find the position of the start and end tokens of our answer in a given feature. We just have to distinguish which parts of the offsets correspond to the question and which part correspond to the context, this is where the `sequence_ids` method of our `tokenized_example` can be useful:

In [23]:
sequence_ids = tokenized_example.sequence_ids()
print(sequence_ids)

[None, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, None, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 

In [24]:
answers = example["answers"]
answers

{'text': ['over 1,600'], 'answer_start': [30]}

### The start value in answer indicates the character offset within the context where the answer starts. We're trying to find the tuple offset of this character in the offset array, thus result of the offset tuple index is less than this value.  Also, we need to find tje feature that contains the full answer. 

In [25]:
ctx_start_index = 0
while sequence_ids[ctx_start_index] != 1:
    ctx_start_index += 1
print(f'context starts at index {ctx_start_index} in sequence_ids')        

context starts at index 16 in sequence_ids


It returns `None` for the special tokens, then 0 or 1 depending on whether the corresponding token comes from the first sentence past (the question) or the second (the context). Now with all of this, we can find the first and last token of the answer in one of our input feature (or if the answer is not in this feature):

In [26]:
answers = example["answers"]
start_char = answers["answer_start"][0]
end_char = start_char + len(answers["text"][0])
print(f'start_char= {start_char} \n end_char={end_char} ')

# find start of context in sequence_ids which is first occurance of 1.
ctx_start = 0
while sequence_ids[ctx_start] != 1:
    ctx_start += 1

# End token index of the current span in the text.
ctx_end = len(tokenized_example["input_ids"][0]) - 1
while sequence_ids[ctx_end] != 1:
    ctx_end -= 1

print(f'context range in sequence: {ctx_start} ~ {ctx_end} ')

# Detect if the answer is out of the span (in which case this feature is labeled with the CLS index).
offsets = tokenized_example["offset_mapping"][0]
if (offsets[ctx_start][0] <= start_char and offsets[ctx_end][1] >= end_char):
    # Move the token_start_index and token_end_index to the two ends of the answer.
    # Note: we could go after the last offset if the answer is the last word (edge case).
    while ctx_start < len(offsets) and offsets[ctx_start][0] <= start_char:
        ctx_start += 1
    start_position = ctx_start - 1
    
    while offsets[ctx_end][1] >= end_char:
        ctx_end -= 1
    end_position = ctx_end + 1
    #found answer within context in current feature
    print(f'answer from {start_position} ~ {end_position}') 
else:
    print("The answer is not in this feature.")

start_char= 30 
 end_char=40 
context range in sequence: 16 ~ 382 
answer from 23 ~ 26


And we can double check that it is indeed the theoretical answer:

In [27]:
print(tokenizer.decode(tokenized_example["input_ids"][0][start_position: end_position+1]))
print(answers["text"][0])

over 1, 600
over 1,600


Now let's put everything together in one function we will apply to our training set. In the case of impossible answers (the answer is in another feature given by an example with a long context), we set the cls index for both the start and end position. We could also simply discard those examples from the training set if the flag `allow_impossible_answers` is `False`. Since the preprocessing is already complex enough as it is, we've kept is simple for this part.

In [28]:
import numpy as np

def prepare_train_features(examples):
    # Some of the questions have lots of whitespace on the left, which is not useful and will make the
    # truncation of the context fail (the tokenized question will take a lots of space). So we remove that
    # left whitespace
    examples["question"] = [q.lstrip() for q in examples["question"]]

    # Tokenize our examples with truncation and padding, but keep the overflows using a stride. This results
    # in one example possibly giving several features when a context is long, each of those features having a
    # context that overlaps a bit the context of the previous feature.
    tokenized_examples = tokenizer(examples["question"], examples["context"],
                                   truncation="only_second", max_length=max_length,
                                   stride=doc_stride, return_overflowing_tokens=True,
                                   return_offsets_mapping=True, padding="max_length")        

    # print(f'tokenized_examples has keys: {tokenized_examples.keys()}')
    
    # Since one example might give us several features if it has a long context, we need a map from a feature to
    # its corresponding example. This key gives us just that.
    sample_mapping = tokenized_examples.pop("overflow_to_sample_mapping")
    
    # The offset mappings will give us a map from token to character position in the original context. This will
    # help us compute the start_positions and end_positions.
    # offset_mapping has shape [number of features][max_length][2] where last dim is tuple
    offset_mapping = tokenized_examples.pop("offset_mapping")
    
    # print(f"sample_mapping={sample_mapping}, offset_mapping's shape shape={np.array(offset_mapping).shape}")
        
    # add 2 keys to the tokenized_examples dict indicating the answer's positions within
    # the feature spans. 
    tokenized_examples["start_positions"] = []
    tokenized_examples["end_positions"] = []

    for i, offsets in enumerate(offset_mapping):        
        # We will label impossible answers with the index of the CLS token.
        input_ids = tokenized_examples["input_ids"][i]
        cls_index = input_ids.index(tokenizer.cls_token_id)

        # Grab the sequence corresponding to that example (to know what is the context and what is the question).
        sequence_ids = tokenized_examples.sequence_ids(i)

        # One example can give several spans, this is the index of the example containing this span of text.
        sample_index = sample_mapping[i]
        answers = examples["answers"][sample_index] # get the answer for this sample
        # If no answers are given, set the cls_index as answer.
        if len(answers["answer_start"]) == 0:
            tokenized_examples["start_positions"].append(cls_index)
            tokenized_examples["end_positions"].append(cls_index)
        else:
            # Start/end character index of the answer in the text.
            start_char = answers["answer_start"][0]
            end_char = start_char + len(answers["text"][0])

            # Start token index of the current span in the text.
            ctx_start = 0
            while sequence_ids[ctx_start] != 1:
                ctx_start += 1

            # End token index of the current span in the text.
            ctx_end = len(input_ids) - 1
            while sequence_ids[ctx_end] != 1:
                ctx_end -= 1

            # Detect if the answer is out of the span (in which case this feature is labeled with the CLS index).
            if not (offsets[ctx_start][0] <= start_char and offsets[ctx_end][1] >= end_char):
                tokenized_examples["start_positions"].append(cls_index)
                tokenized_examples["end_positions"].append(cls_index)
            else:
                # Otherwise move the token_start_index and token_end_index to the two ends of the answer.
                # Note: we could go after the last offset if the answer is the last word (edge case).
                while ctx_start < len(offsets) and offsets[ctx_start][0] <= start_char:
                    ctx_start += 1 # find the tuple in offsets where 1st value is start_char position of answer
                tokenized_examples["start_positions"].append(ctx_start - 1)
                while offsets[ctx_end][1] >= end_char:
                    ctx_end -= 1
                tokenized_examples["end_positions"].append(ctx_end + 1)

    return tokenized_examples

This function works with one or several examples. In the case of several examples, the tokenizer will return a list of lists for each key:

In [29]:
features = prepare_train_features(datasets['train'][248:250])
print(f'returned feature has keys: {features.keys()} ')

returned feature has keys: dict_keys(['input_ids', 'attention_mask', 'start_positions', 'end_positions']) 


To apply this function on all the sentences (or pairs of sentences) in our dataset, we just use the `map` method of our `dataset` object we created earlier. This will apply the function on all the elements of all the splits in `dataset`, so our training, validation and testing data will be preprocessed in one single command. Since our preprocessing changes the number of samples, we need to remove the old columns when applying it.

In [30]:
print(f'before pre-processing, dataset:\n\n {datasets}')
tokenized_datasets = datasets.map(prepare_train_features, batched=True, remove_columns=datasets["train"].column_names)

before pre-processing, dataset:

 DatasetDict({
    train: Dataset({
        features: ['id', 'title', 'context', 'question', 'answers'],
        num_rows: 87599
    })
    validation: Dataset({
        features: ['id', 'title', 'context', 'question', 'answers'],
        num_rows: 10570
    })
})


100%|██████████| 11/11 [00:02<00:00,  3.75ba/s]


In [31]:
print(f'after pre-processing, tokenized dataset: \n\n {tokenized_datasets} ')

after pre-processing, tokenized dataset: 

 DatasetDict({
    train: Dataset({
        features: ['attention_mask', 'end_positions', 'input_ids', 'start_positions'],
        num_rows: 88524
    })
    validation: Dataset({
        features: ['attention_mask', 'end_positions', 'input_ids', 'start_positions'],
        num_rows: 10784
    })
}) 


Even better, the results are automatically cached by the 🤗 Datasets library to avoid spending time on this step the next time you run your notebook. The 🤗 Datasets library is normally smart enough to detect when the function you pass to map has changed (and thus requires to not use the cache data). For instance, it will properly detect if you change the task in the first cell and rerun the notebook. 🤗 Datasets warns you when it uses cached files, you can pass `load_from_cache_file=False` in the call to `map` to not use the cached files and force the preprocessing to be applied again.

Note that we passed `batched=True` to encode the texts by batches together. This is to leverage the full benefit of the fast tokenizer we loaded earlier, which will use multi-threading to treat the texts in a batch concurrently.

## Fine-tuning the model

Now that our data is ready for training, we can download the pretrained model and fine-tune it. Since our task is question answering, we use the `AutoModelForQuestionAnswering` class. Like with the tokenizer, the `from_pretrained` method will download and cache the model for us:

In [32]:
from transformers import AutoModelForQuestionAnswering, TrainingArguments, Trainer

model = AutoModelForQuestionAnswering.from_pretrained(model_checkpoint)

Some weights of the model checkpoint at distilbert-base-uncased were not used when initializing DistilBertForQuestionAnswering: ['vocab_layer_norm.bias', 'vocab_projector.bias', 'vocab_projector.weight', 'vocab_transform.bias', 'vocab_transform.weight', 'vocab_layer_norm.weight']
- This IS expected if you are initializing DistilBertForQuestionAnswering from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing DistilBertForQuestionAnswering from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of DistilBertForQuestionAnswering were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['qa_outputs.weight', 'qa_outputs.bias']
You should probably TRAIN this mode

The warning is telling us we are throwing away some weights (the `vocab_transform` and `vocab_layer_norm` layers) and randomly initializing some other (the `pre_classifier` and `classifier` layers). This is absolutely normal in this case, because we are removing the head used to pretrain the model on a masked language modeling objective and replacing it with a new head for which we don't have pretrained weights, so the library warns us we should fine-tune this model before using it for inference, which is exactly what we are going to do.

### hyper parameters

In [33]:
import math 

epochs=10
lr=1e-4
totalSteps=math.ceil(datasets['train'].num_rows/batch_size)*epochs

### Setup optimzzer and scheduler

In [34]:
from transformers import AdamW
from transformers import get_scheduler

opt=AdamW(model.parameters(), lr=lr)
scheduler = get_scheduler("cosine",optimizer=opt,num_warmup_steps=totalSteps//2, num_training_steps=totalSteps)

To instantiate a `Trainer`, we will need to define three more things. The most important is the [`TrainingArguments`](https://huggingface.co/transformers/main_classes/trainer.html#transformers.TrainingArguments), which is a class that contains all the attributes to customize the training. It requires one folder name, which will be used to save the checkpoints of the model, and all other arguments are optional:

In [35]:
args = TrainingArguments(
    f"test-squad",
    evaluation_strategy = "epoch",
    # learning_rate=lr,
    per_device_train_batch_size=batch_size,
    per_device_eval_batch_size=batch_size,
    num_train_epochs=epochs,
    weight_decay=0.01,    
)

Here we set the evaluation to be done at the end of each epoch, tweak the learning rate, use the `batch_size` defined at the top of the notebook and customize the number of epochs for training, as well as the weight decay.

The last two arguments are to setup everything so we can push the model to the [Hub](https://huggingface.co/models) at the end of training. Remove the two of them if you didn't follow the installation steps at the top of the notebook, otherwise you can change the value of `push_to_hub_model_id` to something you would prefer.

Then we will need a data collator that will batch our processed examples together, here the default one will work:

In [36]:
from transformers import default_data_collator

data_collator = default_data_collator

We will evaluate our model and compute metrics in the next section (this is a very long operation, so we will only compute the evaluation loss during training).

Then we just need to pass all of this along with our datasets to the `Trainer`:

In [37]:
trainer = Trainer(
    model,
    args,
    train_dataset=tokenized_datasets["train"],
    eval_dataset=tokenized_datasets["validation"],
    data_collator=data_collator,
    optimizers=(opt, scheduler), 
    tokenizer=tokenizer,
)

We can now finetune our model by just calling the `train` method:

In [38]:
import os 
dir="trainer-model"
trainer.train(dir if os.path.exists(dir) else None)

***** Running training *****
  Num examples = 88524
  Num Epochs = 10
  Instantaneous batch size per device = 24
  Total train batch size (w. parallel, distributed & accumulation) = 24
  Gradient Accumulation steps = 1
  Total optimization steps = 36890
  1%|▏         | 500/36890 [07:28<8:25:50,  1.20it/s] Saving model checkpoint to test-squad/checkpoint-500
Configuration saved in test-squad/checkpoint-500/config.json


{'loss': 5.4237, 'learning_rate': 2.7397260273972604e-06, 'epoch': 0.14}


Model weights saved in test-squad/checkpoint-500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-500/special_tokens_map.json
  3%|▎         | 1000/36890 [14:31<8:16:53,  1.20it/s]Saving model checkpoint to test-squad/checkpoint-1000
Configuration saved in test-squad/checkpoint-1000/config.json


{'loss': 3.5798, 'learning_rate': 5.479452054794521e-06, 'epoch': 0.27}


Model weights saved in test-squad/checkpoint-1000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-1000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-1000/special_tokens_map.json
  4%|▍         | 1500/36890 [21:28<8:09:07,  1.21it/s] Saving model checkpoint to test-squad/checkpoint-1500
Configuration saved in test-squad/checkpoint-1500/config.json


{'loss': 2.405, 'learning_rate': 8.21917808219178e-06, 'epoch': 0.41}


Model weights saved in test-squad/checkpoint-1500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-1500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-1500/special_tokens_map.json
  5%|▌         | 2000/36890 [28:22<8:00:54,  1.21it/s] Saving model checkpoint to test-squad/checkpoint-2000
Configuration saved in test-squad/checkpoint-2000/config.json


{'loss': 1.8434, 'learning_rate': 1.0958904109589042e-05, 'epoch': 0.54}


Model weights saved in test-squad/checkpoint-2000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-2000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-2000/special_tokens_map.json
  7%|▋         | 2500/36890 [35:13<7:50:40,  1.22it/s] Saving model checkpoint to test-squad/checkpoint-2500
Configuration saved in test-squad/checkpoint-2500/config.json


{'loss': 1.5961, 'learning_rate': 1.3698630136986302e-05, 'epoch': 0.68}


Model weights saved in test-squad/checkpoint-2500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-2500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-2500/special_tokens_map.json
  8%|▊         | 3000/36890 [42:02<7:20:19,  1.28it/s] Saving model checkpoint to test-squad/checkpoint-3000
Configuration saved in test-squad/checkpoint-3000/config.json


{'loss': 1.5257, 'learning_rate': 1.643835616438356e-05, 'epoch': 0.81}


Model weights saved in test-squad/checkpoint-3000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-3000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-3000/special_tokens_map.json
  9%|▉         | 3500/36890 [48:33<7:13:36,  1.28it/s] Saving model checkpoint to test-squad/checkpoint-3500
Configuration saved in test-squad/checkpoint-3500/config.json


{'loss': 1.4106, 'learning_rate': 1.9178082191780822e-05, 'epoch': 0.95}


Model weights saved in test-squad/checkpoint-3500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-3500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-3500/special_tokens_map.json
 10%|█         | 3689/36890 [51:01<6:10:58,  1.49it/s] ***** Running Evaluation *****
  Num examples = 10784
  Batch size = 24
                                                      
 10%|█         | 3689/36890 [52:58<6:10:58,  1.49it/s]

{'eval_loss': 1.3132668733596802, 'eval_runtime': 116.7432, 'eval_samples_per_second': 92.374, 'eval_steps_per_second': 3.855, 'epoch': 1.0}


 11%|█         | 4000/36890 [57:01<7:07:09,  1.28it/s]  Saving model checkpoint to test-squad/checkpoint-4000
Configuration saved in test-squad/checkpoint-4000/config.json


{'loss': 1.3119, 'learning_rate': 2.1917808219178083e-05, 'epoch': 1.08}


Model weights saved in test-squad/checkpoint-4000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-4000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-4000/special_tokens_map.json
 12%|█▏        | 4500/36890 [1:03:32<7:00:22,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-4500
Configuration saved in test-squad/checkpoint-4500/config.json


{'loss': 1.206, 'learning_rate': 2.4657534246575342e-05, 'epoch': 1.22}


Model weights saved in test-squad/checkpoint-4500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-4500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-4500/special_tokens_map.json
 14%|█▎        | 5000/36890 [1:10:04<6:54:28,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-5000
Configuration saved in test-squad/checkpoint-5000/config.json


{'loss': 1.2114, 'learning_rate': 2.7397260273972603e-05, 'epoch': 1.36}


Model weights saved in test-squad/checkpoint-5000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-5000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-5000/special_tokens_map.json
 15%|█▍        | 5500/36890 [1:16:35<6:47:34,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-5500
Configuration saved in test-squad/checkpoint-5500/config.json


{'loss': 1.1978, 'learning_rate': 3.0136986301369862e-05, 'epoch': 1.49}


Model weights saved in test-squad/checkpoint-5500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-5500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-5500/special_tokens_map.json
 16%|█▋        | 6000/36890 [1:23:07<6:41:11,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-6000
Configuration saved in test-squad/checkpoint-6000/config.json


{'loss': 1.1623, 'learning_rate': 3.287671232876712e-05, 'epoch': 1.63}


Model weights saved in test-squad/checkpoint-6000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-6000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-6000/special_tokens_map.json
 18%|█▊        | 6500/36890 [1:29:38<6:34:42,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-6500
Configuration saved in test-squad/checkpoint-6500/config.json


{'loss': 1.1382, 'learning_rate': 3.561643835616438e-05, 'epoch': 1.76}


Model weights saved in test-squad/checkpoint-6500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-6500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-6500/special_tokens_map.json
 19%|█▉        | 7000/36890 [1:36:09<6:28:10,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-7000
Configuration saved in test-squad/checkpoint-7000/config.json


{'loss': 1.1412, 'learning_rate': 3.8356164383561644e-05, 'epoch': 1.9}


Model weights saved in test-squad/checkpoint-7000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-7000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-7000/special_tokens_map.json
 20%|██        | 7378/36890 [1:41:05<5:30:04,  1.49it/s]***** Running Evaluation *****
  Num examples = 10784
  Batch size = 24
                                                        
 20%|██        | 7378/36890 [1:43:02<5:30:04,  1.49it/s]

{'eval_loss': 1.1713428497314453, 'eval_runtime': 116.7748, 'eval_samples_per_second': 92.349, 'eval_steps_per_second': 3.854, 'epoch': 2.0}


 20%|██        | 7500/36890 [1:44:37<6:21:50,  1.28it/s]  Saving model checkpoint to test-squad/checkpoint-7500
Configuration saved in test-squad/checkpoint-7500/config.json


{'loss': 1.0742, 'learning_rate': 4.1095890410958905e-05, 'epoch': 2.03}


Model weights saved in test-squad/checkpoint-7500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-7500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-7500/special_tokens_map.json
 22%|██▏       | 8000/36890 [1:51:08<6:15:27,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-8000
Configuration saved in test-squad/checkpoint-8000/config.json


{'loss': 0.8881, 'learning_rate': 4.383561643835617e-05, 'epoch': 2.17}


Model weights saved in test-squad/checkpoint-8000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-8000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-8000/special_tokens_map.json
 23%|██▎       | 8500/36890 [1:57:40<6:12:42,  1.27it/s]Saving model checkpoint to test-squad/checkpoint-8500
Configuration saved in test-squad/checkpoint-8500/config.json


{'loss': 0.8916, 'learning_rate': 4.657534246575342e-05, 'epoch': 2.3}


Model weights saved in test-squad/checkpoint-8500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-8500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-8500/special_tokens_map.json
 24%|██▍       | 9000/36890 [2:04:11<6:02:00,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-9000
Configuration saved in test-squad/checkpoint-9000/config.json


{'loss': 0.9141, 'learning_rate': 4.9315068493150684e-05, 'epoch': 2.44}


Model weights saved in test-squad/checkpoint-9000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-9000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-9000/special_tokens_map.json
 26%|██▌       | 9500/36890 [2:10:43<5:56:23,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-9500
Configuration saved in test-squad/checkpoint-9500/config.json


{'loss': 0.9277, 'learning_rate': 5.2054794520547945e-05, 'epoch': 2.58}


Model weights saved in test-squad/checkpoint-9500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-9500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-9500/special_tokens_map.json
 27%|██▋       | 10000/36890 [2:17:14<5:49:23,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-10000
Configuration saved in test-squad/checkpoint-10000/config.json


{'loss': 0.9498, 'learning_rate': 5.479452054794521e-05, 'epoch': 2.71}


Model weights saved in test-squad/checkpoint-10000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-10000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-10000/special_tokens_map.json
 28%|██▊       | 10500/36890 [2:23:45<5:43:22,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-10500
Configuration saved in test-squad/checkpoint-10500/config.json


{'loss': 0.9397, 'learning_rate': 5.753424657534247e-05, 'epoch': 2.85}


Model weights saved in test-squad/checkpoint-10500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-10500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-10500/special_tokens_map.json
 30%|██▉       | 11000/36890 [2:30:33<5:50:20,  1.23it/s]Saving model checkpoint to test-squad/checkpoint-11000
Configuration saved in test-squad/checkpoint-11000/config.json


{'loss': 0.9872, 'learning_rate': 6.0273972602739724e-05, 'epoch': 2.98}


Model weights saved in test-squad/checkpoint-11000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-11000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-11000/special_tokens_map.json
 30%|███       | 11067/36890 [2:31:28<5:00:08,  1.43it/s]***** Running Evaluation *****
  Num examples = 10784
  Batch size = 24
                                                         
 30%|███       | 11067/36890 [2:33:30<5:00:08,  1.43it/s]

{'eval_loss': 1.1991620063781738, 'eval_runtime': 121.4866, 'eval_samples_per_second': 88.767, 'eval_steps_per_second': 3.704, 'epoch': 3.0}


 31%|███       | 11500/36890 [2:39:20<5:43:18,  1.23it/s]  Saving model checkpoint to test-squad/checkpoint-11500
Configuration saved in test-squad/checkpoint-11500/config.json


{'loss': 0.7038, 'learning_rate': 6.301369863013699e-05, 'epoch': 3.12}


Model weights saved in test-squad/checkpoint-11500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-11500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-11500/special_tokens_map.json
 33%|███▎      | 12000/36890 [2:46:07<5:34:45,  1.24it/s]Saving model checkpoint to test-squad/checkpoint-12000
Configuration saved in test-squad/checkpoint-12000/config.json


{'loss': 0.7008, 'learning_rate': 6.575342465753424e-05, 'epoch': 3.25}


Model weights saved in test-squad/checkpoint-12000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-12000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-12000/special_tokens_map.json
 34%|███▍      | 12500/36890 [2:52:52<5:29:15,  1.23it/s]Saving model checkpoint to test-squad/checkpoint-12500
Configuration saved in test-squad/checkpoint-12500/config.json


{'loss': 0.7495, 'learning_rate': 6.84931506849315e-05, 'epoch': 3.39}


Model weights saved in test-squad/checkpoint-12500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-12500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-12500/special_tokens_map.json
 35%|███▌      | 13000/36890 [2:59:31<5:10:10,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-13000
Configuration saved in test-squad/checkpoint-13000/config.json


{'loss': 0.7515, 'learning_rate': 7.123287671232876e-05, 'epoch': 3.52}


Model weights saved in test-squad/checkpoint-13000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-13000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-13000/special_tokens_map.json
 37%|███▋      | 13500/36890 [3:06:02<5:03:34,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-13500
Configuration saved in test-squad/checkpoint-13500/config.json


{'loss': 0.7723, 'learning_rate': 7.397260273972603e-05, 'epoch': 3.66}


Model weights saved in test-squad/checkpoint-13500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-13500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-13500/special_tokens_map.json
 38%|███▊      | 14000/36890 [3:12:34<4:57:02,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-14000
Configuration saved in test-squad/checkpoint-14000/config.json


{'loss': 0.8269, 'learning_rate': 7.671232876712329e-05, 'epoch': 3.8}


Model weights saved in test-squad/checkpoint-14000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-14000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-14000/special_tokens_map.json
 39%|███▉      | 14500/36890 [3:19:05<4:50:17,  1.29it/s]Saving model checkpoint to test-squad/checkpoint-14500
Configuration saved in test-squad/checkpoint-14500/config.json


{'loss': 0.8144, 'learning_rate': 7.945205479452055e-05, 'epoch': 3.93}


Model weights saved in test-squad/checkpoint-14500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-14500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-14500/special_tokens_map.json
 40%|████      | 14756/36890 [3:22:25<4:07:32,  1.49it/s]***** Running Evaluation *****
  Num examples = 10784
  Batch size = 24
                                                         
 40%|████      | 14756/36890 [3:24:22<4:07:32,  1.49it/s]

{'eval_loss': 1.3963444232940674, 'eval_runtime': 116.7111, 'eval_samples_per_second': 92.399, 'eval_steps_per_second': 3.856, 'epoch': 4.0}


 41%|████      | 15000/36890 [3:27:32<4:44:07,  1.28it/s]  Saving model checkpoint to test-squad/checkpoint-15000
Configuration saved in test-squad/checkpoint-15000/config.json


{'loss': 0.7087, 'learning_rate': 8.219178082191781e-05, 'epoch': 4.07}


Model weights saved in test-squad/checkpoint-15000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-15000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-15000/special_tokens_map.json
 42%|████▏     | 15500/36890 [3:34:03<4:38:01,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-15500
Configuration saved in test-squad/checkpoint-15500/config.json


{'loss': 0.5739, 'learning_rate': 8.493150684931507e-05, 'epoch': 4.2}


Model weights saved in test-squad/checkpoint-15500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-15500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-15500/special_tokens_map.json
 43%|████▎     | 16000/36890 [3:40:35<4:31:13,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-16000
Configuration saved in test-squad/checkpoint-16000/config.json


{'loss': 0.6223, 'learning_rate': 8.767123287671233e-05, 'epoch': 4.34}


Model weights saved in test-squad/checkpoint-16000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-16000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-16000/special_tokens_map.json
 45%|████▍     | 16500/36890 [3:47:06<4:24:37,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-16500
Configuration saved in test-squad/checkpoint-16500/config.json


{'loss': 0.6617, 'learning_rate': 9.041095890410958e-05, 'epoch': 4.47}


Model weights saved in test-squad/checkpoint-16500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-16500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-16500/special_tokens_map.json
 46%|████▌     | 17000/36890 [3:53:37<4:17:58,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-17000
Configuration saved in test-squad/checkpoint-17000/config.json


{'loss': 0.6965, 'learning_rate': 9.315068493150684e-05, 'epoch': 4.61}


Model weights saved in test-squad/checkpoint-17000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-17000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-17000/special_tokens_map.json
 47%|████▋     | 17500/36890 [4:00:08<4:11:38,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-17500
Configuration saved in test-squad/checkpoint-17500/config.json


{'loss': 0.7222, 'learning_rate': 9.58904109589041e-05, 'epoch': 4.74}


Model weights saved in test-squad/checkpoint-17500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-17500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-17500/special_tokens_map.json
 49%|████▉     | 18000/36890 [4:06:39<4:04:51,  1.29it/s]Saving model checkpoint to test-squad/checkpoint-18000
Configuration saved in test-squad/checkpoint-18000/config.json


{'loss': 0.7429, 'learning_rate': 9.863013698630137e-05, 'epoch': 4.88}


Model weights saved in test-squad/checkpoint-18000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-18000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-18000/special_tokens_map.json
 50%|█████     | 18445/36890 [4:12:27<3:26:21,  1.49it/s]***** Running Evaluation *****
  Num examples = 10784
  Batch size = 24
                                                         
 50%|█████     | 18445/36890 [4:14:24<3:26:21,  1.49it/s]

{'eval_loss': 1.3399085998535156, 'eval_runtime': 116.7424, 'eval_samples_per_second': 92.374, 'eval_steps_per_second': 3.855, 'epoch': 5.0}


 50%|█████     | 18500/36890 [4:15:07<3:57:05,  1.29it/s]  Saving model checkpoint to test-squad/checkpoint-18500
Configuration saved in test-squad/checkpoint-18500/config.json


{'loss': 0.761, 'learning_rate': 9.995370575511151e-05, 'epoch': 5.01}


Model weights saved in test-squad/checkpoint-18500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-18500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-18500/special_tokens_map.json
 52%|█████▏    | 19000/36890 [4:21:38<3:53:19,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-19000
Configuration saved in test-squad/checkpoint-19000/config.json


{'loss': 0.5262, 'learning_rate': 9.95838659949645e-05, 'epoch': 5.15}


Model weights saved in test-squad/checkpoint-19000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-19000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-19000/special_tokens_map.json
 53%|█████▎    | 19500/36890 [4:28:10<3:45:49,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-19500
Configuration saved in test-squad/checkpoint-19500/config.json


{'loss': 0.5573, 'learning_rate': 9.88469246388591e-05, 'epoch': 5.29}


Model weights saved in test-squad/checkpoint-19500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-19500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-19500/special_tokens_map.json
 54%|█████▍    | 20000/36890 [4:34:41<3:42:11,  1.27it/s]Saving model checkpoint to test-squad/checkpoint-20000
Configuration saved in test-squad/checkpoint-20000/config.json


{'loss': 0.5837, 'learning_rate': 9.774833774276278e-05, 'epoch': 5.42}


Model weights saved in test-squad/checkpoint-20000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-20000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-20000/special_tokens_map.json
 56%|█████▌    | 20500/36890 [4:41:13<3:35:02,  1.27it/s]Saving model checkpoint to test-squad/checkpoint-20500
Configuration saved in test-squad/checkpoint-20500/config.json


{'loss': 0.5915, 'learning_rate': 9.62962388596925e-05, 'epoch': 5.56}


Model weights saved in test-squad/checkpoint-20500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-20500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-20500/special_tokens_map.json
 57%|█████▋    | 21000/36890 [4:47:45<3:27:26,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-21000
Configuration saved in test-squad/checkpoint-21000/config.json


{'loss': 0.617, 'learning_rate': 9.450137882173384e-05, 'epoch': 5.69}


Model weights saved in test-squad/checkpoint-21000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-21000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-21000/special_tokens_map.json
 58%|█████▊    | 21500/36890 [4:54:16<3:20:10,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-21500
Configuration saved in test-squad/checkpoint-21500/config.json


{'loss': 0.6327, 'learning_rate': 9.237704614464156e-05, 'epoch': 5.83}


Model weights saved in test-squad/checkpoint-21500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-21500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-21500/special_tokens_map.json
 60%|█████▉    | 22000/36890 [5:00:48<3:13:19,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-22000
Configuration saved in test-squad/checkpoint-22000/config.json


{'loss': 0.6515, 'learning_rate': 8.993896864431826e-05, 'epoch': 5.96}


Model weights saved in test-squad/checkpoint-22000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-22000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-22000/special_tokens_map.json
 60%|██████    | 22134/36890 [5:02:33<2:45:32,  1.49it/s]***** Running Evaluation *****
  Num examples = 10784
  Batch size = 24
                                                         
 60%|██████    | 22134/36890 [5:04:30<2:45:32,  1.49it/s]

{'eval_loss': 1.5497112274169922, 'eval_runtime': 116.6912, 'eval_samples_per_second': 92.415, 'eval_steps_per_second': 3.856, 'epoch': 6.0}


 61%|██████    | 22500/36890 [5:09:13<3:08:10,  1.27it/s]  Saving model checkpoint to test-squad/checkpoint-22500
Configuration saved in test-squad/checkpoint-22500/config.json


{'loss': 0.4557, 'learning_rate': 8.720519699356804e-05, 'epoch': 6.1}


Model weights saved in test-squad/checkpoint-22500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-22500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-22500/special_tokens_map.json
 62%|██████▏   | 23000/36890 [5:15:45<3:00:17,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-23000
Configuration saved in test-squad/checkpoint-23000/config.json


{'loss': 0.4005, 'learning_rate': 8.419597108123054e-05, 'epoch': 6.23}


Model weights saved in test-squad/checkpoint-23000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-23000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-23000/special_tokens_map.json
 64%|██████▎   | 23500/36890 [5:22:16<2:54:18,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-23500
Configuration saved in test-squad/checkpoint-23500/config.json


{'loss': 0.4324, 'learning_rate': 8.093357016312517e-05, 'epoch': 6.37}


Model weights saved in test-squad/checkpoint-23500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-23500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-23500/special_tokens_map.json
 65%|██████▌   | 24000/36890 [5:28:47<2:47:35,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-24000
Configuration saved in test-squad/checkpoint-24000/config.json


{'loss': 0.4151, 'learning_rate': 7.744214791423596e-05, 'epoch': 6.51}


Model weights saved in test-squad/checkpoint-24000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-24000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-24000/special_tokens_map.json
 66%|██████▋   | 24500/36890 [5:35:18<2:41:14,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-24500
Configuration saved in test-squad/checkpoint-24500/config.json


{'loss': 0.4407, 'learning_rate': 7.374755360335253e-05, 'epoch': 6.64}


Model weights saved in test-squad/checkpoint-24500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-24500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-24500/special_tokens_map.json
 68%|██████▊   | 25000/36890 [5:41:49<2:34:32,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-25000
Configuration saved in test-squad/checkpoint-25000/config.json


{'loss': 0.4374, 'learning_rate': 6.98771407141278e-05, 'epoch': 6.78}


Model weights saved in test-squad/checkpoint-25000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-25000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-25000/special_tokens_map.json
 69%|██████▉   | 25500/36890 [5:48:21<2:28:08,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-25500
Configuration saved in test-squad/checkpoint-25500/config.json


{'loss': 0.4303, 'learning_rate': 6.585956442945532e-05, 'epoch': 6.91}


Model weights saved in test-squad/checkpoint-25500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-25500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-25500/special_tokens_map.json
 70%|███████   | 25823/36890 [5:52:33<2:03:42,  1.49it/s]***** Running Evaluation *****
  Num examples = 10784
  Batch size = 24
                                                         
 70%|███████   | 25823/36890 [5:54:30<2:03:42,  1.49it/s]

{'eval_loss': 1.7412137985229492, 'eval_runtime': 116.7128, 'eval_samples_per_second': 92.398, 'eval_steps_per_second': 3.856, 'epoch': 7.0}


 70%|███████   | 26000/36890 [5:56:47<2:20:23,  1.29it/s]  Saving model checkpoint to test-squad/checkpoint-26000
Configuration saved in test-squad/checkpoint-26000/config.json


{'loss': 0.3531, 'learning_rate': 6.172456947852049e-05, 'epoch': 7.05}


Model weights saved in test-squad/checkpoint-26000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-26000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-26000/special_tokens_map.json
 72%|███████▏  | 26500/36890 [6:03:17<2:14:59,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-26500
Configuration saved in test-squad/checkpoint-26500/config.json


{'loss': 0.2414, 'learning_rate': 5.7502769917232635e-05, 'epoch': 7.18}


Model weights saved in test-squad/checkpoint-26500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-26500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-26500/special_tokens_map.json
 73%|███████▎  | 27000/36890 [6:09:48<2:08:31,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-27000
Configuration saved in test-squad/checkpoint-27000/config.json


{'loss': 0.2426, 'learning_rate': 5.3225422472465824e-05, 'epoch': 7.32}


Model weights saved in test-squad/checkpoint-27000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-27000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-27000/special_tokens_map.json
 75%|███████▍  | 27500/36890 [6:16:19<2:02:06,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-27500
Configuration saved in test-squad/checkpoint-27500/config.json


{'loss': 0.2453, 'learning_rate': 4.892419512818889e-05, 'epoch': 7.45}


Model weights saved in test-squad/checkpoint-27500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-27500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-27500/special_tokens_map.json
 76%|███████▌  | 28000/36890 [6:22:50<1:55:33,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-28000
Configuration saved in test-squad/checkpoint-28000/config.json


{'loss': 0.2435, 'learning_rate': 4.463093266679185e-05, 'epoch': 7.59}


Model weights saved in test-squad/checkpoint-28000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-28000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-28000/special_tokens_map.json
 77%|███████▋  | 28500/36890 [6:29:22<1:48:53,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-28500
Configuration saved in test-squad/checkpoint-28500/config.json


{'loss': 0.2327, 'learning_rate': 4.0377420901458506e-05, 'epoch': 7.73}


Model weights saved in test-squad/checkpoint-28500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-28500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-28500/special_tokens_map.json
 79%|███████▊  | 29000/36890 [6:35:53<1:42:41,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-29000
Configuration saved in test-squad/checkpoint-29000/config.json


{'loss': 0.2347, 'learning_rate': 3.619515134512656e-05, 'epoch': 7.86}


Model weights saved in test-squad/checkpoint-29000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-29000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-29000/special_tokens_map.json
 80%|███████▉  | 29500/36890 [6:42:38<1:39:06,  1.24it/s]Saving model checkpoint to test-squad/checkpoint-29500
Configuration saved in test-squad/checkpoint-29500/config.json


{'loss': 0.2229, 'learning_rate': 3.2115088058343725e-05, 'epoch': 8.0}


Model weights saved in test-squad/checkpoint-29500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-29500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-29500/special_tokens_map.json
 80%|████████  | 29512/36890 [6:42:49<1:26:08,  1.43it/s]***** Running Evaluation *****
  Num examples = 10784
  Batch size = 24
                                                         
 80%|████████  | 29512/36890 [6:44:51<1:26:08,  1.43it/s]

{'eval_loss': 2.1368227005004883, 'eval_runtime': 122.4974, 'eval_samples_per_second': 88.035, 'eval_steps_per_second': 3.674, 'epoch': 8.0}


 81%|████████▏ | 30000/36890 [6:51:24<1:31:53,  1.25it/s] Saving model checkpoint to test-squad/checkpoint-30000
Configuration saved in test-squad/checkpoint-30000/config.json


{'loss': 0.113, 'learning_rate': 2.8167438402196805e-05, 'epoch': 8.13}


Model weights saved in test-squad/checkpoint-30000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-30000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-30000/special_tokens_map.json
 83%|████████▎ | 30500/36890 [6:58:09<1:25:42,  1.24it/s]Saving model checkpoint to test-squad/checkpoint-30500
Configuration saved in test-squad/checkpoint-30500/config.json


{'loss': 0.1174, 'learning_rate': 2.438142939357882e-05, 'epoch': 8.27}


Model weights saved in test-squad/checkpoint-30500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-30500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-30500/special_tokens_map.json
 84%|████████▍ | 31000/36890 [7:04:54<1:19:14,  1.24it/s]Saving model checkpoint to test-squad/checkpoint-31000
Configuration saved in test-squad/checkpoint-31000/config.json


{'loss': 0.1076, 'learning_rate': 2.0785091318581577e-05, 'epoch': 8.4}


Model weights saved in test-squad/checkpoint-31000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-31000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-31000/special_tokens_map.json
 85%|████████▌ | 31500/36890 [7:11:35<1:10:03,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-31500
Configuration saved in test-squad/checkpoint-31500/config.json


{'loss': 0.1076, 'learning_rate': 1.7405050206064373e-05, 'epoch': 8.54}


Model weights saved in test-squad/checkpoint-31500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-31500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-31500/special_tokens_map.json
 87%|████████▋ | 32000/36890 [7:18:05<1:03:33,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-32000
Configuration saved in test-squad/checkpoint-32000/config.json


{'loss': 0.1118, 'learning_rate': 1.4266330697851954e-05, 'epoch': 8.67}


Model weights saved in test-squad/checkpoint-32000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-32000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-32000/special_tokens_map.json
 88%|████████▊ | 32500/36890 [7:24:36<57:04,  1.28it/s]  Saving model checkpoint to test-squad/checkpoint-32500
Configuration saved in test-squad/checkpoint-32500/config.json


{'loss': 0.1043, 'learning_rate': 1.1392170775041788e-05, 'epoch': 8.81}


Model weights saved in test-squad/checkpoint-32500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-32500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-32500/special_tokens_map.json
 89%|████████▉ | 33000/36890 [7:31:07<50:29,  1.28it/s]  Saving model checkpoint to test-squad/checkpoint-33000
Configuration saved in test-squad/checkpoint-33000/config.json


{'loss': 0.1023, 'learning_rate': 8.803849712122292e-06, 'epoch': 8.95}


Model weights saved in test-squad/checkpoint-33000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-33000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-33000/special_tokens_map.json
 90%|█████████ | 33201/36890 [7:33:44<41:16,  1.49it/s]  ***** Running Evaluation *****
  Num examples = 10784
  Batch size = 24
                                                       
 90%|█████████ | 33201/36890 [7:35:41<41:16,  1.49it/s]

{'eval_loss': 2.610574960708618, 'eval_runtime': 116.6713, 'eval_samples_per_second': 92.431, 'eval_steps_per_second': 3.857, 'epoch': 9.0}


 91%|█████████ | 33500/36890 [7:39:32<44:01,  1.28it/s]   Saving model checkpoint to test-squad/checkpoint-33500
Configuration saved in test-squad/checkpoint-33500/config.json


{'loss': 0.0687, 'learning_rate': 6.52053053266945e-06, 'epoch': 9.08}


Model weights saved in test-squad/checkpoint-33500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-33500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-33500/special_tokens_map.json
 92%|█████████▏| 34000/36890 [7:46:03<37:32,  1.28it/s]  Saving model checkpoint to test-squad/checkpoint-34000
Configuration saved in test-squad/checkpoint-34000/config.json


{'loss': 0.0476, 'learning_rate': 4.559118133024853e-06, 'epoch': 9.22}


Model weights saved in test-squad/checkpoint-34000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-34000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-34000/special_tokens_map.json
 94%|█████████▎| 34500/36890 [7:52:34<31:02,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-34500
Configuration saved in test-squad/checkpoint-34500/config.json


{'loss': 0.0472, 'learning_rate': 2.934134124357646e-06, 'epoch': 9.35}


Model weights saved in test-squad/checkpoint-34500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-34500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-34500/special_tokens_map.json
 95%|█████████▍| 35000/36890 [7:59:05<24:32,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-35000
Configuration saved in test-squad/checkpoint-35000/config.json


{'loss': 0.0486, 'learning_rate': 1.6576093197361253e-06, 'epoch': 9.49}


Model weights saved in test-squad/checkpoint-35000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-35000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-35000/special_tokens_map.json
 96%|█████████▌| 35500/36890 [8:05:36<18:02,  1.28it/s]Saving model checkpoint to test-squad/checkpoint-35500
Configuration saved in test-squad/checkpoint-35500/config.json


{'loss': 0.0493, 'learning_rate': 7.389946621969679e-07, 'epoch': 9.62}


Model weights saved in test-squad/checkpoint-35500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-35500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-35500/special_tokens_map.json
 98%|█████████▊| 36000/36890 [8:12:14<12:28,  1.19it/s]Saving model checkpoint to test-squad/checkpoint-36000
Configuration saved in test-squad/checkpoint-36000/config.json


{'loss': 0.0526, 'learning_rate': 1.850912532696092e-07, 'epoch': 9.76}


Model weights saved in test-squad/checkpoint-36000/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-36000/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-36000/special_tokens_map.json
 99%|█████████▉| 36500/36890 [8:19:06<05:31,  1.18it/s]Saving model checkpoint to test-squad/checkpoint-36500
Configuration saved in test-squad/checkpoint-36500/config.json


{'loss': 0.0508, 'learning_rate': 0.0, 'epoch': 9.89}


Model weights saved in test-squad/checkpoint-36500/pytorch_model.bin
tokenizer config file saved in test-squad/checkpoint-36500/tokenizer_config.json
Special tokens file saved in test-squad/checkpoint-36500/special_tokens_map.json
100%|██████████| 36890/36890 [8:24:24<00:00,  1.44it/s]***** Running Evaluation *****
  Num examples = 10784
  Batch size = 24
                                                       
100%|██████████| 36890/36890 [8:26:25<00:00,  1.44it/s]

Training completed. Do not forget to share your model on huggingface.co/models =)


100%|██████████| 36890/36890 [8:26:25<00:00,  1.21it/s]

{'eval_loss': 3.0912227630615234, 'eval_runtime': 120.7042, 'eval_samples_per_second': 89.342, 'eval_steps_per_second': 3.728, 'epoch': 10.0}
{'train_runtime': 30385.1617, 'train_samples_per_second': 29.134, 'train_steps_per_second': 1.214, 'train_loss': 0.7439544039193968, 'epoch': 10.0}


TrainOutput(global_step=36890, training_loss=0.7439544039193968, metrics={'train_runtime': 30385.1617, 'train_samples_per_second': 29.134, 'train_steps_per_second': 1.214, 'train_loss': 0.7439544039193968, 'epoch': 10.0})

Since this training is particularly long, let's save the model just in case we need to restart.

In [39]:
trainer.save_model(dir)

Saving model checkpoint to trainer-model
Configuration saved in trainer-model/config.json
Model weights saved in trainer-model/pytorch_model.bin
tokenizer config file saved in trainer-model/tokenizer_config.json
Special tokens file saved in trainer-model/special_tokens_map.json


In [40]:
model.save_pretrained('saved-QA-model')

Configuration saved in saved-QA-model/config.json
Model weights saved in saved-QA-model/pytorch_model.bin


## Evaluation

In [41]:
from transformers import AutoModelForQuestionAnswering

model=AutoModelForQuestionAnswering.from_pretrained('saved-QA-model')

loading configuration file saved-QA-model/config.json
Model config DistilBertConfig {
  "_name_or_path": "distilbert-base-uncased",
  "activation": "gelu",
  "architectures": [
    "DistilBertForQuestionAnswering"
  ],
  "attention_dropout": 0.1,
  "dim": 768,
  "dropout": 0.1,
  "hidden_dim": 3072,
  "initializer_range": 0.02,
  "max_position_embeddings": 512,
  "model_type": "distilbert",
  "n_heads": 12,
  "n_layers": 6,
  "pad_token_id": 0,
  "qa_dropout": 0.1,
  "seq_classif_dropout": 0.2,
  "sinusoidal_pos_embds": false,
  "tie_weights_": true,
  "torch_dtype": "float32",
  "transformers_version": "4.9.2",
  "vocab_size": 30522
}

loading weights file saved-QA-model/pytorch_model.bin
All model checkpoint weights were used when initializing DistilBertForQuestionAnswering.

All the weights of DistilBertForQuestionAnswering were initialized from the model checkpoint at saved-QA-model.
If your task is similar to the task the model of the checkpoint was trained on, you can already use

Evaluating our model will require a bit more work, as we will need to map the predictions of our model back to parts of the context. The model itself predicts logits for the start and en position of our answers: if we take a batch from our validation datalaoder, here is the output our model gives us:

In [42]:
import torch

for batch in trainer.get_eval_dataloader():
    break
batch = {k: v.to(trainer.args.device) for k, v in batch.items()}
with torch.no_grad():
    output = trainer.model(**batch)
output.keys()

odict_keys(['loss', 'start_logits', 'end_logits'])

The output of the model is a dict-like object that contains the loss (since we provided labels), the start and end logits. We won't need the loss for our predictions, let's have a look at the logits:

In [43]:
output.start_logits.shape, output.end_logits.shape

(torch.Size([24, 384]), torch.Size([24, 384]))

We have one logit for each feature and each token. The most obvious thing to predict an answer for each featyre is to take the index for the maximum of the start logits as a start position and the index of the maximum of the end logits as an end position.

In [44]:
output.start_logits.argmax(dim=-1), output.end_logits.argmax(dim=-1)

(tensor([ 46,  57,  78,  43, 113, 162,  72,  35, 162,  34,  73,  41,  80,  86,
         156,  35,  83,  86,  80,  58,  77,  74,  42,  53], device='cuda:0'),
 tensor([ 47,  58,  81,  44, 113, 165,  75,  37, 165,  36,  76,  42,  83,  87,
         158,  35,  83,  87,  83,  60,  80,  74,  43,  54], device='cuda:0'))

This will work great in a lot of cases, but what if this prediction gives us something impossible: the start position could be greater than the end position, or point to a span of text in the question instead of the answer. In that case, we might want to look at the second best prediction to see if it gives a possible answer and select that instead.

However, picking the second best answer is not as easy as picking the best one: is it the second best index in the start logits with the best index in the end logits? Or the best index in the start logits with the second best index in the end logits? And if that second best answer is not possible either, it gets even trickier for the third best answer.


To classify our answers, we will use the score obtained by adding the start and end logits. We won't try to order all the possible answers and limit ourselves to with a hyper-parameter we call `n_best_size`. We'll pick the best indices in the start and end logits and gather all the answers this predicts. After checking if each one is valid, we will sort them by their score and keep the best one. Here is how we would do this on the first feature in the batch:

In [45]:
n_best_size = 20

In [46]:
import numpy as np

start_logits = output.start_logits[0].cpu().numpy()
end_logits = output.end_logits[0].cpu().numpy()
# Gather the indices the best start/end logits:
start_indexes = np.argsort(start_logits)[-1 : -n_best_size - 1 : -1].tolist()
end_indexes = np.argsort(end_logits)[-1 : -n_best_size - 1 : -1].tolist()
valid_answers = []
for start_index in start_indexes:
    for end_index in end_indexes:
        if start_index <= end_index: # We need to refine that test to check the answer is inside the context
            valid_answers.append(
                {
                    "score": start_logits[start_index] + end_logits[end_index],
                    "text": "" # We need to find a way to get back the original substring corresponding to the answer in the context
                }
            )

And then we can sort the `valid_answers` according to their `score` and only keep the best one. The only point left is how to check a given span is inside the context (and not the question) and how to get back the text inside. To do this, we need to add two things to our validation features:
- the ID of the example that generated the feature (since each example can generate several features, as seen before);
- the offset mapping that will give us a map from token indices to character positions in the context.

That's why we will re-process the validation set with the following function, slightly different from `prepare_train_features`:

In [47]:
def prepare_validation_features(examples):
    # Some of the questions have lots of whitespace on the left, which is not useful and will make the
    # truncation of the context fail (the tokenized question will take a lots of space). So we remove that
    # left whitespace
    examples["question"] = [q.lstrip() for q in examples["question"]]

    # Tokenize our examples with truncation and maybe padding, but keep the overflows using a stride. This results
    # in one example possible giving several features when a context is long, each of those features having a
    # context that overlaps a bit the context of the previous feature.
    tokenized_examples = tokenizer(
        examples["question"],
        examples["context"],
        truncation="only_second",
        max_length=max_length,
        stride=doc_stride,
        return_overflowing_tokens=True,
        return_offsets_mapping=True,
        padding="max_length",
    )

    # Since one example might give us several features if it has a long context, we need a map from a feature to
    # its corresponding example. This key gives us just that.
    sample_mapping = tokenized_examples.pop("overflow_to_sample_mapping")

    # We keep the example_id that gave us this feature and we will store the offset mappings.
    tokenized_examples["example_id"] = []

    for i in range(len(tokenized_examples["input_ids"])):
        # Grab the sequence corresponding to that example (to know what is the context and what is the question).
        sequence_ids = tokenized_examples.sequence_ids(i)
        context_index = 1

        # One example can give several spans, this is the index of the example containing this span of text.
        sample_index = sample_mapping[i]
        tokenized_examples["example_id"].append(examples["id"][sample_index])

        # Set to None the offset_mapping that are not part of the context so it's easy to determine if a token
        # position is part of the context or not.
        tokenized_examples["offset_mapping"][i] = [
            (o if sequence_ids[k] == context_index else None)
            for k, o in enumerate(tokenized_examples["offset_mapping"][i])
        ]

    return tokenized_examples

And like before, we can apply that function to our validation set easily:

In [48]:
validation_features = datasets["validation"].map(
    prepare_validation_features,
    batched=True,
    remove_columns=datasets["validation"].column_names
)

100%|██████████| 11/11 [00:11<00:00,  1.02s/ba]


Now we can grab the predictions for all features by using the `Trainer.predict` method:

In [49]:
raw_predictions = trainer.predict(validation_features)

The following columns in the test set  don't have a corresponding argument in `DistilBertForQuestionAnswering.forward` and have been ignored: offset_mapping, example_id.
***** Running Prediction *****
  Num examples = 10784
  Batch size = 24
100%|█████████▉| 449/450 [02:03<00:00,  3.59it/s]

The `Trainer` *hides* the columns that are not used by the model (here `example_id` and `offset_mapping` which we will need for our post-processing), so we set them back:

In [50]:
validation_features.set_format(type=validation_features.format["type"], columns=list(validation_features.features.keys()))

We can now refine the test we had before: since we set `None` in the offset mappings when it corresponds to a part of the question, it's easy to check if an answer is fully inside the context. We also eliminate very long answers from our considerations (with an hyper-parameter we can tune)

In [51]:
max_answer_length = 30

In [52]:
start_logits = output.start_logits[0].cpu().numpy()
end_logits = output.end_logits[0].cpu().numpy()
offset_mapping = validation_features[0]["offset_mapping"]
# The first feature comes from the first example. For the more general case, we will need to be match the example_id to
# an example index
context = datasets["validation"][0]["context"]

# Gather the indices the best start/end logits:
start_indexes = np.argsort(start_logits)[-1 : -n_best_size - 1 : -1].tolist()
end_indexes = np.argsort(end_logits)[-1 : -n_best_size - 1 : -1].tolist()
valid_answers = []
for start_index in start_indexes:
    for end_index in end_indexes:
        # Don't consider out-of-scope answers, either because the indices are out of bounds or correspond
        # to part of the input_ids that are not in the context.
        if (
            start_index >= len(offset_mapping)
            or end_index >= len(offset_mapping)
            or offset_mapping[start_index] is None
            or offset_mapping[end_index] is None
        ):
            continue
        # Don't consider answers with a length that is either < 0 or > max_answer_length.
        if end_index < start_index or end_index - start_index + 1 > max_answer_length:
            continue
        if start_index <= end_index: # We need to refine that test to check the answer is inside the context
            start_char = offset_mapping[start_index][0]
            end_char = offset_mapping[end_index][1]
            valid_answers.append(
                {
                    "score": start_logits[start_index] + end_logits[end_index],
                    "text": context[start_char: end_char]
                }
            )

valid_answers = sorted(valid_answers, key=lambda x: x["score"], reverse=True)[:n_best_size]
valid_answers

[{'score': 25.307564, 'text': 'Denver Broncos'},
 {'score': 23.239866,
  'text': 'Denver Broncos defeated the National Football Conference (NFC) champion Carolina Panthers'},
 {'score': 20.345716, 'text': 'Carolina Panthers'},
 {'score': 13.423489, 'text': 'Broncos'},
 {'score': 11.495771,
  'text': 'The American Football Conference (AFC) champion Denver Broncos'},
 {'score': 11.35579,
  'text': 'Broncos defeated the National Football Conference (NFC) champion Carolina Panthers'},
 {'score': 9.6571045,
  'text': 'National Football Conference (NFC) champion Carolina Panthers'},
 {'score': 9.428073,
  'text': 'The American Football Conference (AFC) champion Denver Broncos defeated the National Football Conference (NFC) champion Carolina Panthers'},
 {'score': 8.819131, 'text': 'Panthers'},
 {'score': 8.265104, 'text': 'Denver'},
 {'score': 6.986588,
  'text': 'Denver Broncos defeated the National Football Conference (NFC) champion Carolina Panthers 24–10'},
 {'score': 6.2996073,
  'text'

We can compare to the actual ground-truth answer:

In [53]:
datasets["validation"][0]["answers"]

{'text': ['Denver Broncos', 'Denver Broncos', 'Denver Broncos'],
 'answer_start': [177, 177, 177]}

Our model picked the right as the most likely answer!

As we mentioned in the code above, this was easy on the first feature because we knew it comes from the first example. For the other features, we will need a map between examples and their corresponding features. Also, since one example can give several features, we will need to gather together all the answers in all the features generated by a given example, then pick the best one. The following code builds a map from example index to its corresponding features indices:

In [54]:
import collections

examples = datasets["validation"]
features = validation_features

example_id_to_index = {k: i for i, k in enumerate(examples["id"])}
features_per_example = collections.defaultdict(list)
for i, feature in enumerate(features):
    features_per_example[example_id_to_index[feature["example_id"]]].append(i)

We're almost ready for our post-processing function. The last bit to deal with is the impossible answer (when `squad_v2 = True`). The code above only keeps answers that are inside the context, we need to also grab the score for the impossible answer (which has start and end indices corresponding to the index of the CLS token). When one example gives several features, we have to predict the impossible answer when all the features give a high score to the impossible answer (since one feature could predict the impossible answer just because the answer isn't in the part of the context it has access too), which is why the score of the impossible answer for one example is the *minimum* of the scores for the impossible answer in each feature generated by the example.

We then predict the impossible answer when that score is greater than the score of the best non-impossible answer. All combined together, this gives us this post-processing function:

In [55]:
from tqdm.auto import tqdm

def postprocess_qa_predictions(examples, features, raw_predictions, n_best_size = 20, max_answer_length = 30):
    all_start_logits, all_end_logits = raw_predictions
    # Build a map example to its corresponding features.
    example_id_to_index = {k: i for i, k in enumerate(examples["id"])}
    features_per_example = collections.defaultdict(list)
    for i, feature in enumerate(features):
        features_per_example[example_id_to_index[feature["example_id"]]].append(i)

    # The dictionaries we have to fill.
    predictions = collections.OrderedDict()

    # Logging.
    print(f"Post-processing {len(examples)} example predictions split into {len(features)} features.")

    # Let's loop over all the examples!
    for example_index, example in enumerate(tqdm(examples)):
        # Those are the indices of the features associated to the current example.
        feature_indices = features_per_example[example_index]

        min_null_score = None # Only used if squad_v2 is True.
        valid_answers = []
        
        context = example["context"]
        # Looping through all the features associated to the current example.
        for feature_index in feature_indices:
            # We grab the predictions of the model for this feature.
            start_logits = all_start_logits[feature_index]
            end_logits = all_end_logits[feature_index]
            # This is what will allow us to map some the positions in our logits to span of texts in the original
            # context.
            offset_mapping = features[feature_index]["offset_mapping"]

            # Update minimum null prediction.
            cls_index = features[feature_index]["input_ids"].index(tokenizer.cls_token_id)
            feature_null_score = start_logits[cls_index] + end_logits[cls_index]
            if min_null_score is None or min_null_score < feature_null_score:
                min_null_score = feature_null_score

            # Go through all possibilities for the `n_best_size` greater start and end logits.
            start_indexes = np.argsort(start_logits)[-1 : -n_best_size - 1 : -1].tolist()
            end_indexes = np.argsort(end_logits)[-1 : -n_best_size - 1 : -1].tolist()
            for start_index in start_indexes:
                for end_index in end_indexes:
                    # Don't consider out-of-scope answers, either because the indices are out of bounds or correspond
                    # to part of the input_ids that are not in the context.
                    if (
                        start_index >= len(offset_mapping)
                        or end_index >= len(offset_mapping)
                        or offset_mapping[start_index] is None
                        or offset_mapping[end_index] is None
                    ):
                        continue
                    # Don't consider answers with a length that is either < 0 or > max_answer_length.
                    if end_index < start_index or end_index - start_index + 1 > max_answer_length:
                        continue

                    start_char = offset_mapping[start_index][0]
                    end_char = offset_mapping[end_index][1]
                    valid_answers.append(
                        {
                            "score": start_logits[start_index] + end_logits[end_index],
                            "text": context[start_char: end_char]
                        }
                    )
        
        if len(valid_answers) > 0:
            best_answer = sorted(valid_answers, key=lambda x: x["score"], reverse=True)[0]
        else:
            # In the very rare edge case we have not a single non-null prediction, we create a fake prediction to avoid
            # failure.
            best_answer = {"text": "", "score": 0.0}
        
        # Let's pick our final answer: the best one or the null answer (only for squad_v2)
        if not squad_v2:
            predictions[example["id"]] = best_answer["text"]
        else:
            answer = best_answer["text"] if best_answer["score"] > min_null_score else ""
            predictions[example["id"]] = answer

    return predictions

And we can apply our post-processing function to our raw predictions:

In [56]:
final_predictions = postprocess_qa_predictions(datasets["validation"], validation_features, raw_predictions.predictions)

Post-processing 10570 example predictions split into 10784 features.


100%|██████████| 10570/10570 [00:23<00:00, 448.26it/s]


Then we can load the metric from the datasets library.

In [57]:
metric = load_metric("squad_v2" if squad_v2 else "squad")

Downloading: 4.51kB [00:00, 1.59MB/s]                   
Downloading: 3.31kB [00:00, 1.43MB/s]                   


Then we can call compute on it. We just need to format predictions and labels a bit as it expects a list of dictionaries and not one big dictionary. In the case of squad_v2, we also have to set a `no_answer_probability` argument (which we set to 0.0 here as we have already set the answer to empty if we picked it).

In [58]:
if squad_v2:
    formatted_predictions = [{"id": k, "prediction_text": v, "no_answer_probability": 0.0} for k, v in final_predictions.items()]
else:
    formatted_predictions = [{"id": k, "prediction_text": v} for k, v in final_predictions.items()]
references = [{"id": ex["id"], "answers": ex["answers"]} for ex in datasets["validation"]]
metric.compute(predictions=formatted_predictions, references=references)

{'exact_match': 73.28287606433302, 'f1': 83.00790512160691}

### Save the optimized model

In [59]:
model.save_pretrained('saved-QA-model')

Configuration saved in saved-QA-model/config.json
Model weights saved in saved-QA-model/pytorch_model.bin
